draft-thomson-http-binary-message.md

```diff
@@ -29,21 +29,594 @@ This document defines a binary format for representing HTTP messages.
 
 # Introduction
 
-Words...
+This document defines a simple format for representing an HTTP message
+({{!HTTP=I-D.ietf-httpbis-semantics}}), either request or response. This allows
+for the encoding of HTTP messages that can be conveyed outside of an HTTP
+protocol. This enables the application of transformations to entire messages,
+including the application of encryption and authentication.
+
+This format is informed by the framing structure of HTTP/2 ({{?H2=RFC7540}})
+and HTTP/3 ({{?H3=I-D.ietf-quic-http}}). In comparison, this format simpler by
+virtue of not including either header compression ({{?HPACK=RFC7541}},
+{{?QPACK=I-D.ietf-quic-qpack}}) or a generic framing layer.
+
+This format provides an alternative to the `message/http` content type defined
+in {{?MESSAGING=I-D.ietf-httpbis-messaging}}. A binary format permits more
+efficient encoding and processing of messages. A binary format also reduces
+exposure to security problems related to processing of HTTP messages.
+
+Two modes for encoding are described:
+
+* a known-length encoding includes length prefixes for all major message
+  components; and
+
+* an indefinite-length encoding enables efficient generation of messages where
+  lengths are not known when starting an encoding.
+
 
 # Conventions and Definitions
 
 {::boilerplate bcp14}
 
+This document uses terminology from HTTP ({{!HTTP}}).
+
+This document uses notation from QUIC ({{!QUIC=I-D.ietf-quic-transport}}).
+
+
+# Format
+
+An HTTP message is split into five sections, following the structure defined in
+Section 6 of {{!HTTP}}:
+
+1. Framing indicator. This format uses a single integer to describe framing, which describes
+   whether the message is a request or response and how subsequent sections are
+   formatted; see {{framing}}
+
+2. Control data. For a request, this contains the request method and target.
+   For a response, this contains the status code.
+
+3. Header section.  This contains zero or more header fields.
+
+4. Content.  This is a sequence of zero or more bytes.
+
+5. Trailer section.  This contains zero or more trailer fields.
+
+A response message can include multiple informational responses, which causes
+the control data and header section to be repeated.
+
+All lengths and numeric values are encoded using the variable-length integer
+encoding from {{!QUIC}}.
+
+
+## Known Length Messages
+
+A message that has a known length at the time of construction uses the
+format shown in {{format-known-length}}.
+
+~~~ quic-format
+Message with Known-Length {
+  Framing (i) = 0..1,
+  Known-Length Informational Response (..) ...,
```

Comment:
What's the meaning of the extra `...` on this line? That it is optional?

Reply:
It means the field is repeated one or more times (I stumbled across this myself).

Reply:
zero or more, yeah.